src/core/libmaven/isotopeDetection.cpp

```diff
@@ -63,207 +63,182 @@ void IsotopeDetection::pullIsotopes(PeakGroup* parentgroup)
-
-            float mzmin = isotopeMass -_mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
-            float mzmax = isotopeMass +_mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
-
-            float rt = parentgroup->medianRt();
-            float rtmin = parentgroup->minRt;
-            float rtmax = parentgroup->maxRt;
-
-            Peak* parentPeak = parentgroup->getPeak(sample);
-            if (parentPeak) {
-                rt = parentPeak->rt;
-                rtmin = parentPeak->rtmin;
-                rtmax = parentPeak->rtmax;
-            }
-
-            float isotopePeakIntensity = 0;
-            float parentPeakIntensity = 0;
-
-            if (parentPeak) {
-                parentPeakIntensity = parentPeak->peakIntensity;
-                Scan* scan = parentPeak->getScan();
-                std::pair<float, float> isotope = getIntensity(scan, mzmin, mzmax);
-                isotopePeakIntensity = isotope.first;
-                rt = isotope.second;
-            }
+    map<string, PeakGroup> isotopeGroups;
+    for (auto sample : _mavenParameters->samples) {
+        for (Isotope& isotope : masslist) {
+            string isotopeName = isotope.name;
+            float expectedAbundance = static_cast<float>(isotope.abundance);
+
+            float isotopeMass = static_cast<float>(isotope.mass);
+            float mzDelta = _mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
+            float mzmin = isotopeMass - mzDelta;
+            float mzmax = isotopeMass + mzDelta;
+
+            Peak* parentPeak = parentGroup->getPeak(sample);
+            // we should not attempt to find isotopic peaks for samples in
+            // which there is no parent peak
+            if (parentPeak == nullptr)
+                continue;
 
-            if (isotopePeakIntensity == 0 || rt == 0) continue;
+            // even though we will do a peak detection on the isotope's slice
+            // later on, this is a good fast check to see if there is anything
+            // worth searching for; maybe also for adduct & fragment detection?
+            Scan* scan = parentPeak->getScan();
+            std::pair<float, float> topMatch = getIntensity(scan, mzmin, mzmax);
+            float isotopePeakIntensity = topMatch.first;
+            float rt = topMatch.second;
 
-            if (filterIsotope(x, isotopePeakIntensity, parentPeakIntensity, sample, parentgroup))
+            if (isotopePeakIntensity == 0.0f || rt == 0.0f)
                 continue;
 
-            vector<Peak> allPeaks;
-
-            EIC * eic = sample->getEIC(mzmin, mzmax, sample->minRt,sample->maxRt, 1, _mavenParameters->eicType,
-                                        _mavenParameters->filterline);
-            //actually last parameter should probably be deepest MS level?
-            //TODO: decide how isotope children should even work in MS mode
+            float rtmin = parentGroup->getSlice().rtmin;
```

Comment:
I would expect the rtmin and max to be the parent peak's rtmin and max, instead of the group range to prevent any issues from minor RT drift across samples. It might also reduce the number of peaks we get per EIC since the range will be narrower. Let me know what you think.

Reply:
Yea, agree. Will change.